### ISSUE TITLE
tls failed to take full effect

### ISSUE THREAD

**Issue by pinghe** [caddyserver/caddy#3683]
2.2.0-rc.1

```

{
  debug
  admin   off
  auto_https off
}
localhost:7180 {

  root * webroot/site0/
  file_server
}
localhost:7181 {

  root * webroot/site1/
  file_server
}

localhost:7182 {

  root * webroot/site2/
  file_server
}


```
The tls attribute is not false.

```
2020/08/25 00:35:02.542	DEBUG	http	starting server loop	{"address": "[::]:7180", "http3": false, "tls": false}
2020/08/25 00:35:02.543	DEBUG	http	starting server loop	{"address": "[::]:7181", "http3": false, "tls": true}
2020/08/25 00:35:02.543	DEBUG	http	starting server loop	{"address": "[::]:7182", "http3": false, "tls": true}
```

**Comment by francislavoie** [comment]
Hmm. You have an extra set of `{ }` around your config. Where did that come from?

I recommend using the `caddy fmt` command to format your Caddyfile, it makes it easier to follow.

If you use `caddy adapt --pretty` to look at the underlying JSON config, you'll see that each of your sites have TLS connection policies. This is added by the Caddyfile adapter. I'll get back to this part later.

So, that log line is printed from here:

https://github.com/caddyserver/caddy/blob/744d04c2585d50f64cf3d43d139c310a18e78f73/modules/caddyhttp/app.go#L374

You'll notice that `useTLS` is set here:

https://github.com/caddyserver/caddy/blob/744d04c2585d50f64cf3d43d139c310a18e78f73/modules/caddyhttp/app.go#L326

Basically, your servers have TLS connection policies due to the Caddyfile adapter, so Caddy is defaulting to TLS.

The simple fix for you for now, is to specify `http://` for each of your sites to ensure that Caddy won't add the connection policies.

That said, I'm thinking that if `auto_https off` is set, the Caddyfile adapter should avoid setting that default connection policy. The code for that is here:

https://github.com/caddyserver/caddy/blob/744d04c2585d50f64cf3d43d139c310a18e78f73/caddyconfig/httpcaddyfile/httptype.go#L579-L583

Specifically, `localhost` as the hostname will make it pass for `addressQualifiesForTLS`, then `autoHTTPSWillAddConnPolicy` is false because of `auto_https off`, which results in the connection policy being added. That doesn't seem exactly right to me.

But I'll need to let @mholt chime in here once he has the time.

**Comment by pinghe** [comment]
Thanks.

```
{
  debug
  admin   off
 # auto_https off
}

http://localhost:7180 {
  root * webroot/site0/
  file_server
}

http://localhost:7181 {

  root * webroot/site1/
  file_server
}

http://localhost:7182 {

  root * webroot/site2/
  file_server
}

http://localhost:7183 {

  root * webroot/site3/
  file_server
}
```

```
2020/08/25 09:27:51.663	DEBUG	http	starting server loop	{"address": "[::]:7180", "http3": false, "tls": false}
2020/08/25 09:27:51.663	DEBUG	http	starting server loop	{"address": "[::]:7181", "http3": false, "tls": false}
2020/08/25 09:27:51.663	DEBUG	http	starting server loop	{"address": "[::]:7182", "http3": false, "tls": false}
2020/08/25 09:27:51.663	DEBUG	http	starting server loop	{"address": "[::]:7183", "http3": false, "tls": false}

```

**Comment by mholt** [comment]
Thanks for the report; this is the offending line: https://github.com/caddyserver/caddy/blob/24f34780b61fc82df6507cf95535c3dfc9f78058/caddyconfig/httpcaddyfile/httptype.go#L567-L583

Specifically, the ` !autoHTTPSWillAddConnPolicy` condition triggers it in this case.

**Comment by mholt** [comment]
For clarity, it's important to understand what **automatic HTTPS** (`auto_https`) is and does. Francis is on the right track. [From the docs](https://caddyserver.com/docs/automatic-https#effects):

> When automatic HTTPS is activated, the following occurs:
>
> - Certificates are obtained and renewed for all domain names
> - The default port (if any) is changed to the HTTPS port 443
> - HTTP is redirected to HTTPS (this uses HTTP port 80)

It does seem odd that a connection policy is being _added_ unnecessarily when `auto_https off` is used.

One way to fix this would be to change `&&` to `||` here: https://github.com/caddyserver/caddy/blob/24f34780b61fc82df6507cf95535c3dfc9f78058/caddyconfig/httpcaddyfile/httptype.go#L478

In other words, one key from the block is sufficient to add a connection policy, rather than needing all of them to add one.

But I don't know if that is correct. Would have to test it against inputs like from #3004 as per the comment.

@pinghe Can you try making the change from && to || and see if that works for you? It looks like it should be making the intended JSON then, in a few quick trials I did.

**Comment by mholt** [comment]
@pinghe Did that work for you? It'd be nice to get a fix in the upcoming 2.2 release but I need you to verify, since you're the one who reported it.

**Comment by mholt** [comment]
Closing due to inactivity.
